modules/livestore/partition_reader.go

```diff
@@ -0,0 +1,307 @@
+package livestore
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/grafana/tempo/pkg/ingest"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+type record struct {
+	tenantID string
+	content  []byte
+	offset   int64
+}
+
+type consumeFn func(context.Context, []record) error
+
+type PartitionReader struct {
+	services.Service
+
+	partitionID   int32
+	consumerGroup string
+	topic         string
+
+	client *kgo.Client
+	adm    *kadm.Client
+
+	consume consumeFn
+	metrics partitionReaderMetrics
+
+	logger log.Logger
+
+	// Watermark-based commit management
+	highWatermark  atomic.Int64
+	commitInterval time.Duration
+	wg             sync.WaitGroup
+}
+
+func NewPartitionReaderForPusher(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newPartitionReaderMetrics(partitionID, reg)
+	return newPartitionReader(client, partitionID, cfg, consume, logger, metrics)
+}
+
+func newPartitionReader(client *kgo.Client, partitionID int32, cfg ingest.KafkaConfig, consume consumeFn, logger log.Logger, metrics partitionReaderMetrics) (*PartitionReader, error) {
+	r := &PartitionReader{
+		partitionID:    partitionID,
+		consumerGroup:  cfg.ConsumerGroup,
+		topic:          cfg.Topic,
+		client:         client,
+		adm:            kadm.NewClient(client),
+		consume:        consume,
+		metrics:        metrics,
+		logger:         log.With(logger, "partition", partitionID),
+		commitInterval: 10 * time.Second, // Commit every 10 seconds
+	}
+
+	r.Service = services.NewBasicService(r.start, r.running, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(context.Context) error {
+	return nil
+}
+
+func (r *PartitionReader) running(ctx context.Context) error {
+	consumeCtx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	offset, err := r.fetchLastCommittedOffsetWithRetries(consumeCtx)
+	if err != nil {
+		return fmt.Errorf("failed to fetch last committed offset: %w", err)
+	}
+	r.client.AddConsumePartitions(map[string]map[int32]kgo.Offset{r.topic: {r.partitionID: offset}})
```

Comment:
I hate this map in map but looks like its what the library expects.

Reply:
Yes